### ISSUE TITLE
Change `__str__` requirement for identity

### ISSUE THREAD

**Issue by luhn** [Pylons/pyramid#3520]
Opening this based on @merwok's https://github.com/Pylons/pyramid/pull/3465#discussion_r330175629

> Hi! Sorry to comment after this is merged, but I was following the issue and not aware of this PR.
> 
> I don’t understand the reason for using `__str__` as the method to convert Identity to user ID.
> 
>     * `__str__` is meant to control what happens with `print`, `%s` formatting, etc.
> 
>     * IMO it is good practice to have `__str__` fast, without side effects, without possibility of failure (so no parsing, casting, etc)
> 
>     * `repr` uses `__str__` if the class doesn’t define `__repr__`
> 
>     * REPL, logging, others use repr or str, so seeing something that’s a user ID instead of custom str/repr output (like `<SomeIdentityClass id=blah source=value>`) in debug logs, etc. could make inspection harder
> 
>     * depending on the system, user ID may not be a string
> 
>     * calling `str(identity)` instead of `identity.some_explicit_method()` seems a bit obscure (not discoverable)
> 
> 
> If people agree with the disadvantages here, could this API decision be reconsidered?

Some previous discussions:

https://github.com/Pylons/pyramid/issues/3422#issuecomment-443840114
https://github.com/Pylons/pyramid/pull/3465#issuecomment-511140572

**Comment by luhn** [comment]
The more I think about it, the more I think @merwok's complaint has merit.

My current favorite idea is adding a `userid()` method to the security policy.  This has a couple advantages:

- The identity can now be entirely opaque.  That was the original intention, but we strayed from that with the requirement of `__str__`.
- In the common case that userid is stored in the session and the identity object stored in a database, the userid can be fetched without needing to query the database.
- Clearer boundaries for the `SessionAuthenticationHelper`, etc.  Helpers implement remember, forget, userid.  The user implements identify and permits.

**Comment by luhn** [comment]
I realize my second and third points don't address the idea of `authenticated_userid` being the userid after it's been checked against the database.

Personally I've never found the unauthenticated_userid/authenticated_userid dichotomy to be useful.  However, it does seem to be important.

Perhaps un-deprecate `unauthenticated_userid`, which naively returns the userid provided by the session or whatever, then has `authenticated_userid` return the userid *only if the identity is not null*.

**Comment by luhn** [comment]
Another idea that a `userid()` method opens up is caching the identity on the request, busting the cache anytime the userid changes.  I frequently have to jury-rig a cache into my security policies, so this would solve a huge pain-point for me.

**Comment by mmerickel** [comment]
The simplest and most direct solution is probably to just get rid of the object right? Pyramid isn't actually using it for anything anymore.

```python
class ISecurityPolicy:
    def authenticated_userid(request):
        ...

    def permits(request, context, permission):
        ...

    def remember(request, userid, **kw):
        ...

    def forget(request, **kw):
        ...
```

The makes the policy actually pretty consistent in that it just focuses on a userid and principals etc are all hidden in permits from the framework's perspective.

**Comment by mmerickel** [comment]
I think the caching situation is huge, and I think we should try to solve it. I'm kind of thinking that if we did, I'm not sure how much of it would/should belong in the security policy anyway or if we could keep that stateless-ish like it always has been. Currently, in my proposal above, I think the main thing missing is any opinion or support for loading that opaque identity object which you want to share between (at least) permits and authenticated_userid.

**Comment by luhn** [comment]
Agreed on everything.   Removing is easy, the lack of identity for `permits` is the downside.  Not the end of the world, since `permits` has the request object and can get whatever it needs, but makes for slightly awkward implementations.

**Comment by luhn** [comment]
To clarify on my caching suggestion, security policy would remain stateless.  I imagined an implementation like this:

```python
class SecurityAPIMixin(object):
    _cached_identity = None
    _cached_identity_userid = None

    @property
    def unauthenticated_userid(self):
        policy = _get_security_policy(self)
        if policy is None:
            return None
        return security.unauthenticated_userid(self)

    @property
    def authenticated_identity(self):
        userid = self.unauthenticated_userid
        if userid is None:
             return None
        if self._cached_identity is not None and userid == self._cached_identity_userid:
            return self._cached_identity
        policy = _get_security_policy(self)
        if policy is None:
            return None
        identity = policy.identify(self, userid)
        self._cached_identity = identity
        self._cached_identity_userid = userid
        return identity

    @property
    def authenticated_userid(self):
        if self.authenticated_identity is not None:
            return self.unauthenticated_userid
        else:
            return None
```

**Comment by mmerickel** [comment]
I'm not sure I'm following the goal of bringing back unauthenticated_userid.

**Comment by luhn** [comment]
I thought people were reluctant to give that up?  I'm very much happy not to bring that back.

You can ignore that part then, just look at the caching in authenticated_identity.

**Comment by mmerickel** [comment]
Yeah unauthenticated_userid is no longer necessary, it was a way of exposing basically the unvalidated value that comes from the super class. Now you just get it directly from the helper.

**Comment by merwok** [comment]
It seems that this went from a small renaming issue to a broader discussion, but that’s good!  Trying to follow and willing to make (or start) a PR.

[mmerickel]
> The simplest and most direct solution is probably to just get rid of the object right? Pyramid isn't actually using it for anything anymore.

Sorry, what object?

[luhn]
> Another idea that a `userid()` method opens up is caching the identity on the request

In projects I’ve seen, caching in auth policy is not an issue because they define a custom reified request property to get a user object (database model), or to get parsed claims and principals from e.g. JWT (in cases with external authentication and no database model).  No need to cache authenticated_userid (old system) when you have cached `request.authenticated_user` :slightly_smiling_face:

**Comment by luhn** [comment]
> Sorry, what object?

The `identity` object from the new security policy.

> No need to cache authenticated_userid (old system) when you have cached request.authenticated_user

Yeah, no need to cache `authenticated_userid`, but if the `identity` object is sticking around, it's definitely desirable to cache that.

**Comment by merwok** [comment]
Should I try changing identify → authenticated_userid ? :smile:

**Comment by luhn** [comment]
I think we should get a few more people to weigh in on the discussion.  @dstufft @stevepiercy @bertjwregeer @ztane 

Summary of discussion as I understand it:

- Top idea is to bring back `authenticated_userid`.  No opposition so far.
- If we do bring back `authenticated_userid`, do we need `identity`?
    - Argument to remove:  Pyramid doesn't use it anywhere.
    - Argument to keep:  Without `identity`, Pyramid would lack "any opinion or support for loading that opaque identity object which you want to share between (at least) permits and authenticated_userid."

**Comment by mmerickel** [comment]
I think it could simply be `policy.authenticated_userid(request, identity)`, right? `request` only being there for convenience / consistency.

**Comment by mmerickel** [comment]
Ok the identity situation is making me lose sleep so I'm hijacking this thread.

1) I think we need an `authenticated_userid(request)` function on the policy.

2) The identity object itself is now, and has always been the constant issue with Pyramid's api. Do we cache it? Do we keep it as is, computed per call? WHAT DO WE DO? It's definitely convenient for applications to have an object to use, whether it is cached or not. Also, personally, I've never written a Pyramid app that didn't cache it.

Well, my big idea, is to come up with a way to let the policy control it. My proposal here is that the return value of `identify` means nothing to Pyramid and is solely there for use by the application and the policy. A bw-compat policy would return `None` here and just re-invoke unauthenticated_userid or effective_principals etc on the legacy policies directly inside of `authenticated_userid` and `permits`. Since the policy is intended to be defined by the application, it should know what the `authenticated_identity` is and what it can do with it.

I really think this api strikes a nice compromise and gives the policy the appropriate control to do what it wants.

```python
class Request:
    @reify
    def authenticated_identity(self):
        # this is reified so anyone can overwrite the value
        # (like setting it to None during logout) and it is only called once
        return policy.identify(self)

class ISecurityPolicy(Interface):
    def identify(request):
        """ Return an object identifying a trusted and verified user.  This
        object may be anything and will be stored as
        :attr:`pyramid.request.Request.authenticated_identity`.
        """

    def authenticated_userid(request):
        """ Return a :term:`userid` string identifying the trusted and
        verified user, or `None` if unauthenticated.
        """

    def permits(request, context, permission):
        """ Return an instance of :class:`pyramid.security.Allowed` if a user
        of the given request is allowed the ``permission`` in the current
        ``context``, else return an instance of
        :class:`pyramid.security.Denied`.
        """

    def remember(request, userid, **kw):
        """ Return a set of headers suitable for 'remembering' the
        :term:`userid` named ``userid`` when set in a response.  An
        individual authentication policy and its consumers can
        decide on the composition and meaning of ``**kw``.
        """

    def forget(request, **kw):
        """ Return a set of headers suitable for 'forgetting' the
        current user on subsequent requests. An individual authentication
        policy and its consumers can decide on the composition and
        meaning of ``**kw``.
        """
```

**Comment by luhn** [comment]
Ha, I've lost some sleep over this security policy as well.

We're definitely on the same page here—I agree with everything you're saying and think its a good way forward.

I feel like an exception is the best course for both a legacy policy and a policy without identify defined.  Since identity is only meaningful to the application, it should be an error if any code tries to use it if it hasn't been defined.

What do you think of my idea having identify have `authenticated_userid` as an argument, then busting the cache if the authenticated_userid?  That addresses some concerns brought up a long time ago that caching could be dangerous because the authenticated user could change mid-request (e.g. login or logout pages).

```python
class SecurityAPIMixin(object):
    _cached_identity = None
    _cached_identity_userid = None

    @property
    def authenticated_identity(self):
        userid = self.unauthenticated_userid
        if userid is None:
             return None
        if self._cached_identity is not None and userid == self._cached_identity_userid:
            return self._cached_identity
        policy = _get_security_policy(self)
        if policy is None:
            return None
        identity = policy.identify(self, userid)
        self._cached_identity = identity
        self._cached_identity_userid = userid
        return identity

    @property
    def authenticated_userid(self):
        policy = _get_security_policy(self)
        if policy is None:
            return None
        return security.unauthenticated_userid(self)
```

**Comment by mmerickel** [comment]
@luhn I think that is implementable inside the policy itself in user code using my proposed features because the policy can clear or modify `request.authenticated_identity` in `remember`/`forget` if it wants to do that.

A possibly cleaner api that I dismissed because of implementation issues (but I'll bring up because it's solvable and the optional identify method is a little bit of a wart), is to get rid of authenticated_identity entirely and have `identify` not support a return value. So `identify` could mutate the request object itself directly however it wants to and the other methods can use that value. The caveat is that you only want to call `identify` once per request and the implementation detail is we'd need a way to enforce that. But I don't see it as a dealbreaker and it would be cleaner to say "modify the request however you want". This would allow the policy to manage `request.user` or something else if it wanted and it'd document what attributes it managed on the request per-policy (by default managing nothing).

**Comment by mmerickel** [comment]
I wonder if this is reason enough to turn the security policy into a per-request instance. Imagine `request.authenticator = security_factory(request)` - basically the same workflow as `request.session`. Then it can store attrs on `self` during the request lifecycle if it chooses. The factory basically replaces `identify` and the other methods stay the same, without the request argument. Just need to come up with a name for the thing. Sorry to keep bikeshedding this but I think this discussion is really getting things to a better place so far.

**Comment by luhn** [comment]
> So identify could mutate the request object itself directly however it wants to and the other methods can use that value.

To me this seems very wrong.  It solves the problem, but I can't think of part of Pyramid that operates this way.  Plus the implementation issues you've pointed out, seems ripe for bugs.

I do like the idea of per-request identity.  It gives full control to the developer for caching, lazy/eager loading, etc.

If I understand you right, you're suggesting something like this?  (Ignoring the naming)

```python
class IdentityFactory:
    def __call__(self, request):
        # Load stuff, if desired.
        return Identity(request)

class Identity:
    def __init__(self, request):
        self.request = request

    def authenticated_userid(self):
        pass  # Could also be a property?

    def permits(self, context, principal):
        pass

   def remember(self, userid, **kw):
        pass


   def forget(self, userid):
        pass

    # User can define whatever additional methods and properties they want
    # accessible via `request.identity.attr`


class Request:
    @reify
    def identity(self):
        factory = get_identity_factory()
        return factory(self)
```

Initial impression is positive.  But the second thought is "permits, remember, and forget would make more sense to be outside the factory (especially remember)"  Which then brings us full circle to a non-opaque identity object...

**Comment by mmerickel** [comment]
> But the second thought is "permits, remember, and forget would make more sense to be inside the factory (especially remember)" Which then brings us full circle to a non-opaque identity object...

My goal by keeping them on this cached object is that remember/forget can be used as lifecycle events to invalidate the cached identity, if the `Identity` per-request object wants to do so. I don't really like calling this an identity, but I'm also struggling to come up with a name for it.

**Comment by luhn** [comment]
Okay, I see what you're getting at.  I think that should solve all the problems and gives the user a lot more flexibility.  (There was some [previous discussion](https://github.com/Pylons/pyramid/issues/3422#issuecomment-440841725) on a per-request policy.)  Downside is it conflates the concepts of security policy (who is this user, what can they do?) and identity (what are the attributes of this user?).  Thinking of how that would fit in my applications, or projects like pyramid_multiauth, starts to really confuse me.

Hot, possibly regressive take:  Why do we need a userid at all? The original complaint was "identity shouldn't require __str__ == userid".  We could just drop the requirement entirely and not replace it with anything.  I know you said you wanted it, but:

- Pyramid doesn't care about userid (or identity)
- The userid (and identity) is meaningless for any third-party packages, its only of value to the developer.
- Logging and debugging are better off using `identity` so the user can implement __str__ and __repr__ as they see fit.
- We can still keep `authenticated_userid` and `unauthenticated_userid` around for legacy policies, but won't work for new policies.  Or have them optional in the security policy for those concerned about bw compat.

Downsides:

- `remember` does still have `userid`.  I don't find that problematic.
- Lots of momentum behind authenticated_userid (tutorials, sample code, etc.)

Sorry if I'm backtracking here, but worth reconsidering our assumptions.

I'm still in favor of the per-request policy, regardless.

**Comment by mmerickel** [comment]
Your list of downsides are the reasons I want to keep the userid around. I think they are very compelling reasons and I think remembering a userid is unavoidable in the api - the symmetry of being able to remember it and then get it back later is totally worth it.

I think if we can't come up with a good name for an object (I don't think identity fits due to some reasons you stated above) then I'd go back to my previous proposal [1] which is to define `request.authenticated_identity`. The lack of ability to come up with a good name indicates it'll be confusing for people and not worth doing. It's a relatively minor difference whether the class holds the state or it returns the state and pyramid puts it on the request. One nice thing about the policy route is that we don't give the user direct access to the policy anywhere so we can control it better in pyramid. For example if we went with the object, users could do `request.authenticator.remember(...)` instead of going through `pyramid.security.remember()`. This could be considered more flexible but it also opens up for more abuse / confusion.

[1] https://github.com/Pylons/pyramid/issues/3520#issuecomment-543010738

**Comment by stevepiercy** [comment]
Perhaps "entity"?
 
 

 
 
 
 
>

**Comment by luhn** [comment]
Fair.  I think your original proposition is good:  straightforward and solves the issue.

And we're all onboard with making the policy per-request too?  I'm pretty excited for that change, it'll make policy implementation much easier.

**Comment by mmerickel** [comment]
The original proposal I was linking to doesn't have a per-request policy. Just the cached authenticated_identity that the policy can use, loaded from the optional identify method.

**Comment by luhn** [comment]
The per-request is orthogonal, we can do proposal with or without. I think it would be a huge boon.

**Comment by mmerickel** [comment]
I mean I gave several reasons above why the per-request thing isn’t hashed out well enough yet to be something I’m comfortable with.

**Comment by luhn** [comment]
Oh, sorry, I think we’re talking about different things. I should have been more clear.  I’m talking about passing ‘request’ into ‘__init__’ and removing ‘request’ from the methods. It doesn’t fundamentally change the API.

**Comment by mmerickel** [comment]
I think I disagree that it fundamentally changes the api. You need to store the object somewhere and if it’s on the request then it’s effectively public which totally changes the access patterns users can use. Also the naming just doesn’t work to call it a policy.

**Comment by luhn** [comment]
My bad, I misunderstood the conversation earlier.  I was thinking of the naming in terms of the shape, not the lifecycle of the object.  Which now I can see how that backs us into a corner with the purpose of the object and a name for it.

**Comment by bertjwregeer** [comment]
I'm a big fan of per request for the policy rather than just one that gets passed a request everywhere.

But I've struggled with the fact that it is not per user, and have all kinds of hacks like storing state on the request object in a "hidden" key used only by my policy and "abusing" pyramid_services to be able to have per-request storage.

**Comment by mmerickel** [comment]
I would really like to avoid making the entire api available on the request. Meaning I don’t want it to be easy to do something like `request.authenticator.remember()` or permits or forget. The thing on the request feels like it should just be the identification payload and thus I think the policy with the `identify` method is the way to go. Data returned from there is stored on `request.authenticated_identity` and is available throughout the app as well as in remember/forget/permits. I think this is a happy balance. I don’t think that we even need to document `identify` as optional. If it isn’t defined then attempts to access `request.authenticated_identity` will just fail as they would now. Users with smarter policies would benefit from using it.

**Comment by bertjwregeer** [comment]
I fully agree with the above, I hadn't thought through having it available directly on the request would do. I like the automatic fallback. Make it so!

**Comment by mmerickel** [comment]
I'll add a recap here to talk specifically about how to support caching the identity. There is a TLDR at the bottom if you don't want the stream of consciousness.

First, I think we all agree we want:

- add `authenticated_userid(request) -> str`
- probably drop the `identity` argument from `permits` because it's so easy to just get it via `request.authenticated_identity` already that it's a little redundant and inconsistent with `remember` and `forget` and they may also want to use the current identity.

Finally, how do we opt-in or opt-out of caching the identity?

<details>
    <summary>Click to expand if you want to watch my internal struggle</summary>

So a couple options I see immediately on how to accomplish this:

Option 1 (this was https://github.com/Pylons/pyramid/issues/3520#issuecomment-543010738):

- modify `request.authenticated_identity` to be reified.
- drop the `identity` argument from `permits` - the reason for this is because of the next two points below, we don't want to require anyone except user code to rely on `request.authenticated_identity` because a policy may choose to not support it
- for policies that do not want a cached identity, just do not define `identify(request) -> Any` or have it raise a helpful exception, or return `None`, basically it's up to them to define what they do with `request.authenticated_identity` publicly, and internally they can do whatever they need to do in each of the methods `authenticated_userid`, `remember`, `forget`, `permits` to re-identify the user from current request values
- for policies that do want it cached, define `identify(request) -> Any` and then use the value of `request.authenticated_identity` inside `authenticated_userid`, `remember`, `forget`, `permits`, and user-code can also use the value

Option 2:

- require a policy to declare in some way that it wants the identity cached (or not cached) - maybe via `policy.reify_identity = True` or `identify(request) -> Tuple[Any, bool]` where the second attribute is a bool indicating whether to reify the value or not.
- probably drop the `identity` argument from `permits` because it's so easy to just get it via `request.authenticated_identity` already that it's a little redundant and inconsistent with `remember` and `forget` and they may also want to use the current identity.
- policy and user code can all use `request.authenticated_identity` happily, and when not cached (for example in the legacy policy), then `identify` will just be called per-access instead of once.

I prefer option 2 because in option 1 there's no clear way to provide the identity to user code since `request.authenticated_identity` is intentionally broken when you don't want to cache the value.

TLDR - should caching be opt-in or opt-out? How should a policy declare whether it wants the identity cached or not?
</details>

I think the best way to do is to assume it's not cached (pyramid has a long history with this decision and I see a lot of benefits in the simplicity of keeping it that way) and have the policy cache it if it wants to - basically turning `request.authenticated_identity` into a thin call to `policy.identify` and allowing the policy to do what it wants instead of adding any caching support to pyramid itself:

```python
class SecurityPolicy:
    def identify(self, request):
        identity = getattr(request, '_cached_identity', marker)
        if identity is marker:
            identity = request._cached_identity = load_identity(request)
        return identity
```

Yes, it requires some magic value on the request but that's fine, people can come up with clever solutions to that problem (maybe a private `weakref` lookup table of request -> identity) and it greatly simplifies pyramid to not have to think about it and leave it all up to the policy.

**Comment by mmerickel** [comment]
I'll say also that given the above, I'm happy to go either way on the identity argument to permits. It's fine to leave there as it's obviously needed more in permits than in the other methods. Amusingly after all of this I've basically talked myself back into the current design with the addition of `authenticated_userid(request) -> str` so the original name of this issue works perfectly.....

**Comment by merwok** [comment]
> There are only two changes from master here:

I still think a specific method name would be better than `__str__` :slightly_smiling_face:

**Comment by merwok** [comment]
> add authenticated_userid(request, identity) -> Optional[str]

To replace the method that’s `__str__` at the moment, right?

**Comment by mmerickel** [comment]
> To replace the method that’s __str__ at the moment, right?

That's correct, this issue was originally about a better approach to getting a userid value for pyramid to use than the current `str(request.authenticated_identity)` approach on master branch. We got side tracked pretty hard because everyone wants to cache the authenticated identity, but I hope I've wrapped us back around to where we were on that.

**Comment by bertjwregeer** [comment]
```
    def authenticated_userid(self, request, identity):
        if identity:
            return str(identity.id)
```

where does the second argument (`identity`) come from? Does Pyramid call `identify` before passing it to `authenticated_userid`?

**Comment by mmerickel** [comment]
Yeah same as permits. The framework would call identify for you. I’d be fine with dropping the argument to it and permits as well if people want.

**Comment by luhn** [comment]
I'd prefer no `identity` argument in authenticated_userid, since the dependency could go the other way—Identity could depend on userid.  Imagining my own implementations, I'd pull authenticated_userid directly from the session and then use authenticated_userid in my identify method to load the correct object from the database.

`identity` argument in permits I could take or leave.

**Comment by mmerickel** [comment]
There's benefits in saying you **can't** consider anything "authenticated" until you load the actual identity object. Sure the session says the userid is this, but what if the user was deleted? You haven't checked yet, so should it really be used by anyone? In the old policy the userid in the payload counted as "unauthenticated_userid" (a poor name, but indicating it hadn't been authenticated yet by verifying against a backend or trust store - only from the payload directly). Having pyramid enforce the loading of the identity first is a safer api.

That being said, if you want to drop it from authenticated_userid then drop it from permits as well and the policy can do things in the order it wishes. From an api perspective, keeping the identity in the signature enforces the contract that the policy doesn't determine who the current user is - it's passed in. Without it, the policy is left to decipher it from the request itself via `self.identify()` or `request.authenticated_identity` and it's more up in the air.

Anyway, I've said what I want to say there. Ultimately I'm fine with dropping it from the api if there's consensus to do that.

**Comment by luhn** [comment]
The convention of "authenticated_userid should be checked against the database" has always bothered me because I've never found it relevant.  I never delete my users, I only ever deactivate/ban them, i.e. permits unilaterally returns Deny.

I say drop identity argument and let the user decide what guarantees they want authenticated_userid to provide.  But in the end it's irrelevant to the implementation of my security policies.

**Comment by merwok** [comment]
Working on this, for the moment keeping `identity` param in.

A question after re-reading the various summary messages:

> add **kw to forget

> ```python
>     def forget(self, request, **kw):
>         request.session.invalidate()
>         return self.authn.forget(request, **kw)
> ```

OK to add varkwargs to ISecurityPolicy.forget, but the existing (and now deprecated) IAuthenticationPolicy class doesn’t accept them.  Should I change it too?  IIUC classic authn/authz are kept for compat only and existing implementations are not going to expect or accept other params than `request` so this shouldn’t be changed, right?

**Comment by merwok** [comment]
Question: adding `authenticated_userid` method to ISecurityPolicy means that I should un-deprecate `Request.authenticated_userid` property, agreed?

**Comment by mmerickel** [comment]
> OK to add varkwargs to ISecurityPolicy.forget, but the existing (and now deprecated) IAuthenticationPolicy class doesn’t accept them. Should I change it too?

Do not change the old policy.. the legacy shim will need to just drop the kwargs before calling the `authn.forget`.

> Question: adding authenticated_userid method to ISecurityPolicy means that I should un-deprecate Request.authenticated_userid property, agreed?

Yes, I thought we already had, but it should not be deprecated.

**Comment by mmerickel** [comment]
Also, thank you for working on it!

**Comment by merwok** [comment]
FTR after changing the code and most of the tests docs, I’m +1 too for removing `identity` param from `ISecurityPolicy.unauthenticated_userid`.  It seems better to let each policy decide which of these builds upon the other (or not).

From my projects I see various ways happening easily:

- request header contains bearer token → identify (get session from redis, get user from DB, confirm user is active) → authenticated_userid (return `str(request.authentified_identity.id)`
- request signed cookie contains user ID → authenticated_userid (parses info, checks expiration in metadata) → identify (call authenticated_userid, get user object)
- request contains JWT → identify (parses JWT) / authenticated_userid (parses JWT) (use case: microservice without database, trusts the info in JWT)

PR coming soon and some more notes/questions :slightly_smiling_face:

**Comment by merwok** [comment]
The previous PR added version changed notes in docs and deprecation warnings in code, which is nice and useful for people upgrading to Pyramid 2, but the test suite does not silence the warning for tests that explicitly exercise the old functionality.  (One can see that by using pytest rather than nose runner.)  IMO it’s a best practice to silence expected warnings, otherwise people get trained to ignore them and important notes from third parties or Python are missed.

**Comment by mmerickel** [comment]
@merwok agreed silencing the expected warnings is best practice. Obviously a bit tedious. We need to switch the test suite to pytest at some point!

**Comment by mmerickel** [comment]
@merwok how's the work going on the api changes?

**Comment by merwok** [comment]
I sort of took a break to make a Pyramid app! :smiley: 
One of the goals was to test the new security system, but honestly it was also for the fun of making a full website for once (I do a lot of APIs and libraries).


I will get back to fixing tests and docs for this ticket next week for sure.
Is there a timeline for the next (alpha I suppose) release that will incorporate auth{n,z} → security policy?

**Comment by mmerickel** [comment]
> Is there a timeline for the next (alpha I suppose) release that will incorporate auth{n,z} → security policy?

For me there isn't much left, probably a few hours of stuff but nothing major is holding up a release at this point.

**Comment by mmerickel** [comment]
@merwok Any chance you wanna push what you have so I can finish it up? I'm a little antsy to get this wrapped up.

**Comment by luhn** [comment]
I’m happy to lend a hand if I can be of use. 

— Theron

**Comment by merwok** [comment]
Hi!  I’ve been taken by other things, but will come back to this later this week.  Code and tests are not hard to change, but I found many docs to be updated now that there isn’t authn/authz policies.  Maybe the next alpha can include this change, and another one full docs update.

**Comment by mmerickel** [comment]
@luhn do you want to take this on? otherwise I'll probably work on it this weekend.

@merwork it'd be good if you wanted to at least push what you have - no one wants to ignore your efforts but there's nothing for us to see or use right now.

**Comment by luhn** [comment]
Sure, I should be able to knock it out this weekend.

**Comment by merwok** [comment]
I’ll send what I have.  I wanted to do separate commits for separate things (restore authenticated_userid, change permits, fix old authn/z references in docs, etc) but if I can’t do that I will make one big messy incomplete commit.

**Comment by luhn** [comment]
@merwok, any chance you push your work today?  I'm going on a plane flight later and will be working on this.

**Comment by merwok** [comment]
Yes stay tuned

**Comment by merwok** [comment]
Draft PR is here!  Thanks for the help @luhn, feel free to branch off my branch and send PR(s) to my repo.  I will wait for your next messages / commits before I work more on this, don’t want to spend time debugging the same failing tests.  Maybe I could work on documentation.

Questions or tasks I noted while working on this:

1. Why should userid always be a string?  If the meaning is app-specific, it could be 
2. Method is named `identify`, variable name is `identity`, isn’t there a risk of confusion when reading docs or writing code?
3. One method lets you get a `user ID`, but the other gives you an `identity`, not a `user`, is this consistent?
4. Official tutorials in various spots will need updates to change authn+authz to security policy.

**Comment by luhn** [comment]
Thanks @merwork!  Taking a look now.

1) Isn't the idea of `authenticated_userid` is it's *not* opaque, unlike `identity`?  So I think it should be a string.
2) You know, now that you mention that it has been a mild pain point during the implementation.  Having the function named `identity` would be consistent with `authenticated_userid`.
3) There was a lot of discussion about the naming of `identity`, IIRC `user` was rejected because it was too generic and likely to conflict with existing codebases.

@mmerickel, what are your thoughts?

**Comment by merwok** [comment]
1. I don’t know about opacity of `authenticated_userid`!  Does Pyramid itself do anything with it?

2. Personally I’m not a fan of methods named with nouns rather than verbs.  But I could go for `authenticated_identity`, mirroring the other name (but too long?)

3. Thanks, that makes sense.  I’m remembering a project with custom request properties named `authenticated_user` and `authenticated_administrator` (database objects or None), so that project would keep these even when switching to Pyramid 2 and defining custom security policy with `identity` object (containing only validated request data, not a DB object).  IOW +1 to name the Pyramid concept `identity` only and keep any `user` concept app-specific.

**Comment by luhn** [comment]
1. To clarify, the implementation doesn’t care and will just pass through the value. Pyramid doesn’t use it anywhere.  So it’s only documented as being a string. This is the same state of affairs as the legacy system. 
2. ‘Get_identity’ works for me. Should we do ‘get_authenticated_userid’ for consistency?  Nows the time.

**Comment by mmerickel** [comment]
I would prefer not to rename things to `get_foo` as the old policies don't do it (for example `effective_principals`) and I don't think it's very useful. 

The `authenticated_userid` should always be a string. It is not opaque as its primary purpose is logging but may also be something the user generates for usage with `remember`. We want to be able to use the id with pyramid's `pyramid.debug_authorization` setting as well as in things like the toolbar, or pyramid_tm attaches it to `transaction.user = userid`.

**Comment by luhn** [comment]
👍 

One last thought: `identify` -> `authenticated_identity`.  Consistent with `authenticated_userid`, matches the corresponding `Request.authenticated_userid`.

**Comment by mmerickel** [comment]
I don't think the payload returned by `identify` needs to represent anything specific so calling it authenticated feels a bit wrong to me. It's not like we require it to return `None` if the user is unauthenticated, etc like we do with `authenticated_userid`.

**Comment by luhn** [comment]
But `authenticated_identity` just passes through the result of `identify`, so what makes that "authenticated"?  Does `authenticated_identity` need to check `authenticated_userid is not None` to make sure the user is authenticated?

**Comment by luhn** [comment]
I was reviewing the conversation:

> There's benefits in saying you can't consider anything "authenticated" until you load the actual identity object. Sure the session says the userid is this, but what if the user was deleted? You haven't checked yet, so should it really be used by anyone?

So I'm pretty confused on what guarantees we're expecting from the `identify` vs `request.authenticated_identity` vs `authenticated_userid`.

**Comment by mmerickel** [comment]
Ok so I forgot that `identify` in the policy is mapped to `authenticated_identity` on the request. So we should probably document and expect that if it returns `None` that there is no user. We'd expect `authenticated_identity` and `authenticated_userid` to return consistent values here (something versus `None`). I'll adjust my review accordingly. With that in mind I'm +0 on renaming `identify` to `authenticated_identity` on the security policy.

**Comment by luhn** [comment]
I’m a +.5, so I think I’ll do it. It’s just a search replace and nows the time to change the API. 

— Theron

**Comment by mmerickel** [comment]
My thought process has always been (and I know you disagree here) that `identify` is the entry point and should be invoked first. Now, when we have `authenticated_identity(request)` and `authenticated_userid(request)`, the situation is no longer clear and for a user it makes it a more difficult decision for them which one to implement first. I'd ask that we try to do something to improve that situation - even if it's just the docs of the two methods. It's one of those things that's just hard to get right I think.

**Comment by mmerickel** [comment]
I'm fine with renaming it `authenticated_identity`. That being said - part of me still really wants `authenticated_identity` to be cached, so I'm struggling with that final decision. I think it makes more sense to do it after the changes in this PR because Pyramid actually doesn't need to call it anywhere. The only APIs that use the identity are in the policy itself or in user code - as it's no longer called by Pyramid to pass an identity to the `permits` method - leaving that up to the policy. Does it make sense what I'm saying? Basically the policy could decide if it supports `authenticated_identity` and if it doesn't then it can just not be defined or raise an exception when it's accessed.

**Comment by luhn** [comment]
Yeah, I'll save it for another PR.

I'm +1 on making `authenticated_identity` optional, I think it does makesthings clearer.

Definitely would like to figure out identity caching as well.  I'll open an issue for discussion.  Maybe now that `authenticated_userid` is doing the heavy lifting we'll have more leeway.

**Comment by mmerickel** [comment]
The caching works if we drop `identity` argument from `permits` which I noticed that #3545 is not currently doing but I think it, or another PR, should.

**Comment by luhn** [comment]
I pushed a commit earlier tonight that took it out. 

I can put it in another PR if you like. 

— Theron

**Comment by mmerickel** [comment]
No need for another PR I just didn’t see it. Thanks!

**Comment by merwok** [comment]
> My thought process has always been […] that `identify` is the entry point and should be invoked first.

I like this and it makes things clear!  (the different use cases I noted in https://github.com/Pylons/pyramid/issues/3520#issuecomment-546658898 would still all be easy to implement cleanly with that model)

**Comment by merwok** [comment]
Would you mind clarifying:

> I'm +1 on making `authenticated_identity` optional, I think it does makesthings clearer.

Does this mean make the method optional on security policy (i.e. not required to define method), or the request property (how?), or means that one of these (or both) may return `None`?

> Definitely would like to figure out identity caching as well. I'll open an issue for discussion.

Great!

> Maybe now that `authenticated_userid` is doing the heavy lifting we'll have more leeway.

What does heavy lifting mean here?  And is this referencing the security policy method or the request property?

**Comment by luhn** [comment]
> I'm +1 on making authenticated_identity optional, I think it does makesthings clearer.

Make it optional on the security policy.  TBD on how `request.authenticated_identity` would behave in that case.  Exception?  `None`?

> Maybe now that authenticated_userid is doing the heavy lifting we'll have more leeway.

Currently we have `authenticated_userid` documented with guarantees it should provide (trusted and valid user) and `identify` with no documented guarantees.  Making `identify` optional further solidifies it as an convenience function rather than a core part of the security policy.  So we have to worry less about the security implications of caching `authenticated_identity`, which was a large concern.

**Comment by luhn** [comment]
More discussion on identity cache over at #3549

**Comment by mmerickel** [comment]
> Make it optional on the security policy. TBD on how request.authenticated_identity would behave in that case. Exception? None?

My thought was an implementor would just not define the method which would trigger a standard exception. They could also define it and raise their own.

Pyramid could do a half-hearted getattr to see if it’s defined and raise a useful message if it’s not when the reified `authenticated_identity` attr is accessed on the request.

**Comment by merwok** [comment]
Great to see this done! :rocket: :ribbon: :rainbow:
